File: 0456. 132 Pattern/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

class Solution {
    public boolean find132pattern_MK1(int[] nums) {
        /*
        Time complexity: O(nlgn)
        Space complexity: O(n)
        */
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(nums[nums.length - 1]);
        int maxK = Integer.MIN_VALUE;
        for (int i = nums.length - 2; i >= 0; --i) {
            if (nums[i] < maxK) {
                return true;
            }
            while (!stack.isEmpty() && nums[i] > stack.peek()) {
                maxK = stack.pop();
            }
            stack.push(nums[i]);
        }
        return false;
    }

    public boolean find132pattern_MK2(int[] nums) {
        /*
        Time complexity: O(nlgn)
        Space complexity: O(n)
        */
        Deque<Integer> stack = new ArrayDeque<>();
        List<Integer> preMin = new ArrayList<>();
        preMin.add(Integer.MAX_VALUE);
        for (int i = 0; i < nums.length; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] <= nums[i]) {
                stack.pop();
            }
            if (!stack.isEmpty() && preMin.get(stack.peek()) < nums[i]) {
                return true;
            }
            stack.push(i);
            preMin.add(Math.min(preMin.get(preMin.size() - 1), nums[i]));
        }
        return false;
    }
}

File: 0456. 132 Pattern/Solution.py
from typing import List


class Solution:
    def find132pattern_MK1(self, nums: List[int]) -> bool:
        stack = []
        min_k = float('-inf')
        for i in range(len(nums) - 1, -1, -1):
            while stack and stack[-1] < nums[i]:
                min_k = stack.pop()
            if nums[i] < min_k:
                return True
            stack.append(nums[i])
        return False

    def find132pattern_MK2(self, nums: List[int]) -> bool:
        stack = []
        pre_min = [float('inf')]
        for i, e in enumerate(nums):
            while stack and nums[stack[-1]] <= e:
                stack.pop()
            if stack and pre_min[stack[-1]] < e:
                return True
            stack.append(i)
            pre_min.append(min(pre_min[-1], e))
        return False
